### ISSUE TITLE
Writing nested F# records

### ISSUE THREAD

**Issue by nrandell** [JoshClose/CsvHelper#983]
I've been trying to write F# records where one record is nested within another one.

```
type ScheduleComparison = {
    Stops: int
    StopsBefore: int
    NewStops: int
    RemovedStops: int
    Trips: int
    TripsBefore: int
    NewTrips: int
    RemovedTrips: int
    MissingSchedules: int
}

type AnalysedSchedule = {
    ServiceDay: string
    Route: string
    Direction: string
    Pattern: string
    Comparison: ScheduleComparison
}
```

When I write records out, I only get the first 4 fields written
```
ServiceDay,Route,Direction,Pattern
2017-12-25,106,1,69
```

I've tracked this down to `ClassMap.AutoMapMembers` , specifically line 287

```
var memberTypeInfo = member.MemberType().GetTypeInfo();
var isDefaultConverter = typeConverterType == typeof( DefaultTypeConverter );
if( isDefaultConverter && ( memberTypeInfo.HasParameterlessConstructor() || memberTypeInfo.IsUserDefinedStruct() ) )
```

I can see that it uses the default type converter, but an F# record doesn't have a parameterless constructor, nor is it a user defined struct, so it doesn't then map the members.

However I can make it work by modifying `IsUserDefinedStruct` to test for the presence of `IStructuralComparable`, which F# records implement (see below)

```
public static bool IsUserDefinedStruct( this Type type )
{
	return (type.IsValueType && !type.IsPrimitive && !type.IsEnum) || 
		type.FindInterfaces((i, _) => i == typeof(IStructuralComparable), null).Length == 1;
}
```

Is this the correct approach? It's probably better to modify `AutoMapMembers` to test for `IStructuralComparable` directly rather than modify `IsUserDefinedStruct`, but before I submit a patch, I wanted to check this was a sensible approach?

For reference, here is what an F# record implementation looks like.
```
        [CompilationMapping(SourceConstructFlags.RecordType)]
        public sealed class ScheduleComparison : IEquatable<ScheduleComparison>, IStructuralEquatable, IComparable<ScheduleComparison>, IComparable, IStructuralComparable
        {
            public ScheduleComparison(int stops, int stopsBefore, int newStops, int removedStops, int trips, int tripsBefore, int newTrips, int removedTrips, int missingSchedules);
...
        }
```

**Comment by JoshClose** [comment]
Have you tried using a `ClassMap`?

**Comment by nrandell** [comment]
A class map works ok, but I was trying to be lazy and not use a class map. :-)

I guess it's also possible to add something to the configuration object that can be used to mark the class as a user defined struct, so testing for `IStructuralComparable` doesn't need to be the only solution.

**Comment by JoshClose** [comment]
You could change what `Configuration.ShouldUseConstructorParameters` tests for. That's a property that you can set yourself.

**Comment by nrandell** [comment]
I've done that, but it still didn't write the deconstructed class out. That's ok for reading in, but not writing out.

**Comment by JoshClose** [comment]
This only works with a C# class if you pass the other type into the constructor like this.

```cs
void Main()
{
	using (var stream = new MemoryStream())
	using (var writer = new StreamWriter(stream))
	using (var reader = new StreamReader(stream))
	using (var csv = new CsvReader(reader))
	{
		writer.WriteLine("Id,Name");
		writer.WriteLine("1,one");
		writer.WriteLine("2,two");
		writer.Flush();
		stream.Position = 0;
		
		csv.GetRecords<A>().ToList().Dump();
	}
}

public class A
{
	public int Id { get; set; }
	public B B { get; set; }

	public A(int Id, B B)
	{
		this.Id = Id;
		this.B = B;
	}
}

public class B
{
	public string Name { get; set; }
	
	public B(string Name)
	{
		this.Name = Name;
	}
}
```

Do F# classes have any constructor on them?

**Comment by nrandell** [comment]
Looks like we've both been working on something to demonstrate this :-)

Have a look at https://github.com/nrandell/csvhelper-issue-938

This shows a working C# test, a failing F# test and a failing C# test. 

F# records are basically sealed classes with parameterized constructors. Hopefully the test in the repository is self explanatory.

The question is then what is the best way to solve this problem. I guess adding a new property to the writer configuration that gets tested at the same time as `IsUserDefinedStruct` feels like the correct place to me, but you know the library far better than I do!

**Comment by JoshClose** [comment]
I get this error trying to build.

> The current .NET SDK does not support targeting .NET Core 2.1.  Either target .NET Core 2.0 or lower, or use a version of the .NET SDK that supports .NET Core 2.1.

Trying to figure out why that is. I installed the latest .NET Core SDK.

**Comment by nrandell** [comment]
I've just pushed a change to downgrade to netcoreapp2.0!

That should work for you now

**Comment by JoshClose** [comment]
If I installed 2.1.201 from here https://www.microsoft.com/net/download/windows

Shouldn't that work?

**Comment by nrandell** [comment]
Look at https://www.microsoft.com/net/download/dotnet-core/sdk-2.1.300-preview1

(It's got some really nice functionality for performance with Span<T> and other stuff - looking forward to that because I think it can make the CsvHelper speed up quite a bit...)

**Comment by JoshClose** [comment]
I've read a little bit about it. I'll have to look into it more once it's released.

**Comment by JoshClose** [comment]
Does version 2.1.104 not work with netcoreapp2.1?

**Comment by nrandell** [comment]
It's all to do with the difference between SDK and runtimes. I think the MS guys are a bit embarrassed about the confusion. See https://github.com/dotnet/designs/blob/9a84136/proposals/sdk-version-scheme.md for more info - but basically 2.1.104 is the SDK, not the runtime. 

Hopefully my downgrade to `netcoreapp2.0` should fix it!

**Comment by JoshClose** [comment]
I'm updating VS now, so it'll be a bit. ;)

**Comment by nrandell** [comment]
Good luck - sorry for putting you through this. I'll catch up when I wake up in the morning ...

**Comment by nrandell** [comment]
You probably don't want to read this https://blogs.msdn.microsoft.com/dotnet/2018/04/11/announcing-net-core-2-1-preview-2/
The link I sent you may already be a bit old ...

**Comment by JoshClose** [comment]
It seems to work fine when reading, and just doesn't work when writing. The culprit is this line https://github.com/JoshClose/CsvHelper/blob/644c9402901ba9218b72b1b5adc427f6970b9f81/src/CsvHelper/CsvWriter.cs#L331 and this line https://github.com/JoshClose/CsvHelper/blob/644c9402901ba9218b72b1b5adc427f6970b9f81/src/CsvHelper/Expressions/ObjectRecordWriter.cs#L49

It's only adding the members and since this is using constructor parameters, they don't get written.

**Comment by JoshClose** [comment]
It's actually mapping just fine. It runs through the `AutoMapConstructorParameters` method, so the mapped constructor parameters are there, along with properties. You should see the a collection of parameters mapped also. It basically needs to do what the reader is doing and check if there are parameters mapped. https://github.com/JoshClose/CsvHelper/blob/644c9402901ba9218b72b1b5adc427f6970b9f81/src/CsvHelper/Expressions/ObjectRecordCreator.cs#L45

**Comment by nrandell** [comment]
I couldn't see any work on this bug, so have tried to fix it myself. The pull request fixes it and all the other tests pass ok.

**Comment by JoshClose** [comment]
I haven't had much time to devote to this library lately. I'll try and take a look soon.

**Comment by JoshClose** [comment]
This was fixed in 9.1.0 and is on nuget. Sorry it took so long.
